projects/packages/analyzer/src/diff-generator.php

```diff
@@ -0,0 +1,83 @@
+<?php
+namespace Automattic\Jetpack\Analyzer;
+
+class Scripts {
+	static function get_differences( $new_path, $old_path ) {
+		$excludes = array( '.git', 'vendor', 'tests', 'docker', 'bin', 'scss', 'images', 'docs', 'languages', 'node_modules' );
+
+		$new_out_path             = dirname( __DIR__ ) . '/output/differences/new_target.json';
+		$old_out_path             = dirname( __DIR__ ) . '/output/differences/old_target.json';
+		$diff_path                = dirname( __DIR__ ) . '/output/differences/diff.json';
+		$jetpack_new_declarations = new Declarations();
+		$jetpack_old_declarations = new Declarations();
+		$differences              = new Differences();
+
+		echo "Scanning new declarations\n";
+		if ( file_exists( $new_out_path ) ) {
+			$jetpack_new_declarations->load( $new_out_path );
+		} else {
+			$jetpack_new_declarations->scan( $new_path, $excludes );
+			$jetpack_new_declarations->save_json( $new_out_path, false );
+		}
+
+		echo "Scanning old declarations\n";
+		if ( file_exists( $old_out_path ) ) {
+			$jetpack_old_declarations->load( $old_out_path );
+		} else {
+			$jetpack_old_declarations->scan( $old_path, $excludes );
+			$jetpack_old_declarations->save_json( $old_out_path, false );
+		}
+
+		echo "Looking for differences\n";
+		if ( file_exists( $diff_path ) ) {
+			$differences->load( $diff_path );
+		} else {
+			$differences->find( $jetpack_new_declarations, $jetpack_old_declarations, $new_path );
+			$differences->save_json( $diff_path, false );
+		}
+
+		return $differences;
+	}
+
+	static function get_warnings( $folder_name, $differences, $excludes ) {
+		echo "Looking for invocations in:\n${folder_name}\n\n";
+		$invocations = new Invocations();
+		$invocations->scan( $folder_name, $excludes );
+		$invocations->save( dirname( __DIR__ ) . '/output/invocations/' . basename( $folder_name ) . '.json', false );
+
+		echo "Generate warnings\n";
+		$warnings = new Warnings();
+		$warnings->generate( $invocations, $differences );
+		$warnings->output();
+		$warnings->save_json( dirname( __DIR__ ) . '/output/warnings/' . basename( $folder_name ) . '.json', false );
+	}
+}
+
+class Locker {
+	// public $__lock_file = '/jp-analyzer.pid';
+	public static $__lock_file = 'jp-analyzer.pid';
+
+	static function lock() {
+		if ( ! self::is_locked() ) {
+			file_put_contents( self::lock_file(), getmypid() );
```

Comment:
You've got a race condition here. Consider what happens if two processes are running simultaneously and both pass line 61 before either executes line 62.

Instead of `file_put_contents()`, something like this should work:
```php
$fp = fopen( self::lock_file(), 'x' );
if ( ! $fp ) {
    throw new \Exception( 'Locked already' );
}
fwrite( $fp, (string) getmypid() );
fclose( $fp );
```
Opening the file with the 'x' flag will make the open fail if some other process already managed to create it.

Reply:
Thanks!

Reply:
hm. `x` flag generates the `WARNING` if the file already exists, which is probably not what we want here. Using `@fopen` solves the problem, but probably in a ugly way 😬 :

Reply:
Yeah, that warning should be suppressed. I guess the question is whether we go the easy way with `@` or do something more complicated. :shrug: I don't really have an opinion on that.